File: backend/app/schemas/audit.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict


class AuditEventResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    event_id: str
    occurred_at: datetime
    action: str
    category: str
    outcome: str
    actor_type: str
    request_id: str | None = None
    source_ip_hash: str | None = None
    target_type: str | None = None
    target_id: str | None = None
    task_id: str | None = None
    schedule_id: int | None = None
    summary: str
    changes: dict[str, Any] | None = None
    metadata: dict[str, Any] | None = None
    error_code: str | None = None

    @classmethod
    def from_model(cls, row: Any) -> "AuditEventResponse":
        import json

        changes = None
        if row.changes_json:
            try:
                changes = json.loads(row.changes_json)
            except Exception:
                pass
        metadata = None
        if row.metadata_json:
            try:
                metadata = json.loads(row.metadata_json)
            except Exception:
                pass

        return cls(
            event_id=row.event_id,
            occurred_at=row.occurred_at,
            action=row.action,
            category=row.category,
            outcome=row.outcome,
            actor_type=row.actor_type,
            request_id=row.request_id,
            source_ip_hash=row.source_ip_hash,
            target_type=row.target_type,
            target_id=row.target_id,
            task_id=row.task_id,
            schedule_id=row.schedule_id,
            summary=row.summary,
            changes=changes,
            metadata=metadata,
            error_code=row.error_code,
        )


class AuditLogPage(BaseModel):
    events: list[AuditEventResponse]
    total: int
    limit: int
    offset: int

File: frontend/src/api/audit.ts
import { request, getAuthHeader, apiBase } from './base';

export type JsonValue =
  | string
  | number
  | boolean
  | null
  | JsonValue[]
  | { [key: string]: JsonValue };

export interface AuditEvent {
  event_id: string;
  occurred_at: string;
  action: string;
  category: string;
  outcome: string;
  actor_type: string;
  request_id?: string;
  source_ip_hash?: string;
  target_type?: string;
  target_id?: string;
  task_id?: string;
  schedule_id?: number;
  summary: string;
  changes?: Record<string, { from: JsonValue; to: JsonValue; changed?: boolean }>;
  metadata?: Record<string, JsonValue>;
  error_code?: string;
}

export interface AuditLogPage {
  events: AuditEvent[];
  total: number;
  limit: number;
  offset: number;
}

export interface AuditQueryParams {
  action?: string;
  category?: string;
  outcome?: string;
  actor_type?: string;
  task_id?: string;
  schedule_id?: number;
  date_from?: string;
  date_to?: string;
  limit?: number;
  offset?: number;
}

export function getAuditLogs(params: AuditQueryParams = {}): Promise<AuditLogPage> {
  const query = new URLSearchParams();
  Object.entries(params).forEach(([key, val]) => {
    if (val !== undefined && val !== null && val !== '') {
      query.append(key, String(val));
    }
  });
  const queryString = query.toString();
  const path = `/api/audit${queryString ? `?${queryString}` : ''}`;
  return request<AuditLogPage>(path);
}

export async function downloadAuditExport(params: AuditQueryParams & { format: 'json' | 'csv' }): Promise<void> {
  const query = new URLSearchParams();
  Object.entries(params).forEach(([key, val]) => {
    if (val !== undefined && val !== null && val !== '') {
      query.append(key, String(val));
    }
  });
  const queryString = query.toString();
  const path = `/api/audit/export${queryString ? `?${queryString}` : ''}`;

  const response = await fetch(`${apiBase}${path}`, {
    headers: {
      ...getAuthHeader(),
    },
  });

  if (!response.ok) {
    throw new Error(`Export failed: ${response.status}`);
  }

  const blob = await response.blob();
  const url = window.URL.createObjectURL(blob);
  const a = document.createElement('a');
  a.href = url;
  a.download = `audit-log-${new Date().toISOString().split('T')[0]}.${params.format}`;
  document.body.appendChild(a);
  a.click();
  a.remove();
  window.URL.revokeObjectURL(url);
}
